programs/cleos/main.cpp

```diff
@@ -436,12 +431,52 @@ fc::variant push_transaction( signed_transaction& trx, const std::vector<public_
       if (tx_use_old_rpc) {
          return call(push_txn_func, packed_transaction_v0(trx, compression));
       } else {
-         try {
-            return call(send_txn_func, packed_transaction_v0(trx, compression));
-         }
-         catch (chain::missing_chain_api_plugin_exception &) {
-            std::cerr << "New RPC send_transaction may not be supported. Add flag --use-old-rpc to use old RPC push_transaction instead." << std::endl;
-            throw;
+         if( !amqp_address.empty() ) {
+            eosio::chain::named_thread_pool thread_pool( "amqp", 1 );
+            transaction_trace_msg trace_msg;
+            eosio::transaction_msg msg{packed_transaction( std::move( trx ), true, compression )};
+            auto buf = fc::raw::pack( msg );
+            const auto& tid = msg.get<packed_transaction>().id();
+            const string id = tid.str();
+            {
+               eosio::amqp qp_trx( fc::logger::get( DEFAULT_LOGGER ), thread_pool.get_executor(), amqp_address, "trx" );
+               eosio::amqp qp_trace( fc::logger::get( DEFAULT_LOGGER ), thread_pool.get_executor(), amqp_address,
+                                     "trace" );
```

Comment:
All instances of cleos are reading from the same queue, which means that the messages will only be dispatched to one of them, right?  They'll only be dispatched to another consumer after the first consumer fails to ack them.

Reply:
That's correct. And if no one ACKs that message it keeps getting requeued forever unless you attach ttl rules to the queue, etc.
Really for an RPC use case like cleos, I think the AMQP thing is to use a server named exclusive queue that gets set with `reply-to`

Reply:
Does RabbitMQ have expiration of messages? Could move to a `reply-to`  but I don't think that is the actual use case in that others are likely to be interested in the trace. Thinking about it I'm not sure even if the message should be ack as others might want to see the trace.

Reply:
Dropping acks is not a good solution, because the message will still only be delivered to one client at a time.  It will only be re-queued after you close the connection.

Reply:
If you want multicast, you need a queue per recipient.

Reply:
That can be done in rabbitmq right? Setup a fanout or whatever? So here should we assume we are the only consumer and ack, reject appropriately?

Reply:
We shouldn't assume that we're the only consumer.  As Matt suggested, we should use a server named exclusive queue, which enforces uniqueness.  The fanout is set up by telling the exchange to send the message to multiple queues.  It depends on the exchange type, the message's routing key, and the queue's binding key.  The exchange type determines how the routing key is matched against the binding keys of all the queues.

Reply:
The two main options are to set the routing key to the transaction ID or to set it to a unique ID specified by the client (i.e. reply-to).

Reply:
I suppose that we could set it up to broadcast all transaction traces, but that's kind of wasteful when cleos only cares about one specific transaction.

Reply:
>I don't think that is the actual use case in that others are likely to be interested in the trace

To me, the plugin seems to be trying to do two different use cases: a textbook RPC use case (which makes sense to use reply-to). But then a different use case which is a 1-to-many broadcast (which makes sense to spam everything to a single exchange and let the exchange and bound queues decide their semantics)

I'm not entirely sure I understand the full _ultimate_ use cases though. Like in the 1-to-many broadcast case at that point the transactions aren't in blocks yet, is that deceptive for the use case? Or, say, if you're going to run multiple instances of this plugin all broadcasting, will the downstream listeners understand they can get duplicates.

Reply:
I'm not opposed to creating two plugins. One that publishes all traces to an exchange and another that serves as an RPC endpoint using an explicit reply-to or the current implicit reply-to and correlation-id approach. I thought the current plugin served both those by providing the publish all as an option but it might be cleaner to have them as separate plugins which would allow for different configuration and running them on different nodes.

As for using reply-to, see discussion on Correlation id here: https://www.rabbitmq.com/tutorials/tutorial-six-python.html I was just using an implicit reply-to of "trace". That "trace" could come in via the request if you think that is cleaner.

The original request for this plugin as I understood it was that it would run on the block signing node so that the "speculative execution" would be the actual execution into the block. It would not be run on other nodes.

Reply:
Correlation id is intended to distinguish multiple requests from a single client.  Even that tutorial recommends that each client should have its own queue and use reply-to.

Reply:
It implies that there is overhead in the reply-to approach. RabbitMQ even has a proprietary way of doing reply-to to offset the overhead (https://www.rabbitmq.com/direct-reply-to.html). Also I don't think the use case is for a blocking client. I would imagine different parts of the system are going to be interested in transactions.

Proposal:
- Create a `amqp_trace_plugin` which will do what `amqp-trx-publish-all-traces` does now and nothing more.
- `amqp_trx_plugin` will be incoming only. I don't see a need for a blocking client. cleos will be blocking as it is now, but cleos is the the intended user. I only added support to cleos so it could be easily tested.

Thoughts? If the desire is for a reply-to, I can certainly add it.

Reply:
The overhead of reply-to is not because of reply-to per se, but is rather a consequence of using short-lived clients.  Making cleos listen to the broadcast would impose the same overhead (and more).  With that being said, I'm fine with not returning a trace from `amqp_trx_plugin`.